### PR TITLE
Support CTRL+Click for opening links in attribute table

### PR DESCRIPTION
**Pull request by manisandro**
Allow directly opening links in the attribute table with Ctrl+Click:

image

Fairly limited implementation, only works if attribute value contains exactly one link (or at least something that resembles being a link).

### CODE REVIEW

File: src/gui/attributetable/qgsattributetablemodel.cpp [hunk 1]
```diff
@@ -698,17 +698,37 @@ QVariant QgsAttributeTableModel::data( const QModelIndex &index, int role ) cons
   }
 
   QVariant val = mFeat.attribute( fieldId );
+  bool isLink = false;
+  if ( val.type() == QVariant::String )
+  {
+    QString textVal = val.toString();
+    if ( QRegExp( "^(http|https|ftp|file)://\\S+$" ).indexIn( textVal ) != -1 )
```
Comment (nyalldawson) [inline src/gui/attributetable/qgsattributetablemodel.cpp:705]: Please use QRegularExpression instead -- QRegExp is deprecated and removed in qt 6. 

Also would be good to move this logic to QgsStringUtils -- there's similar methods in there

File: src/core/qgsstringutils.cpp [hunk 2]
```diff
@@ -563,6 +563,11 @@ QString QgsStringUtils::insertLinks( const QString &string, bool *foundLinks )
   return converted;
 }
 
+bool QgsStringUtils::isUrl( const QString &string )
+{
+  return QRegularExpression( "^(http|https|ftp|file)://\\S+$" ).match( string ).hasMatch();
```
Comment (nyalldawson) [inline src/core/qgsstringutils.cpp:568]:
```suggestion
  const thread_local rxUrl QRegularExpression( "^(http|https|ftp|file)://\\S+$" ); 
  return rxUrl.match( string ).hasMatch();
```

Given that this function WILL be called many millions of times in a single QGIS session, it's worth optimising!

File: src/gui/attributetable/qgsattributetablemodel.cpp [hunk 3]
```diff
@@ -698,17 +699,29 @@ QVariant QgsAttributeTableModel::data( const QModelIndex &index, int role ) cons
   }
 
   QVariant val = mFeat.attribute( fieldId );
+  bool isLink = val.type() == QVariant::String && QgsStringUtils::isUrl( val.toString() );
```
Comment (nyalldawson) [inline src/gui/attributetable/qgsattributetablemodel.cpp:702]: Again, given that we'll be evaluating this many millions of times, can you only evaluate this when the role is one which requires this knowledge? (ie tooltip/foreground role/font role)

File: src/core/qgsstringutils.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/qgsstringutils.cpp:568]: (needs unit tests too)

File: tests/src/core/testqgsstringutils.cpp [hunk 4]
```diff
@@ -258,6 +259,15 @@ void TestQgsStringUtils::wordWrap()
...
 }
 
+void TestQgsStringUtils::testIsUrl()
+{
+  QVERIFY( QgsStringUtils::isUrl( QStringLiteral( "http://example.com" ) ) );
+  QVERIFY( QgsStringUtils::isUrl( QStringLiteral( "https://example.com" ) ) );
+  QVERIFY( QgsStringUtils::isUrl( QStringLiteral( "ftp://example.com" ) ) );
+  QVERIFY( QgsStringUtils::isUrl( QStringLiteral( "file:///path/to/file" ) ) );
+  QVERIFY( QgsStringUtils::isUrl( QStringLiteral( "file://C:\\path\\to\\file" ) ) );
+}
```
Comment (nyalldawson) [inline tests/src/core/testqgsstringutils.cpp:269]: Can you add some non urls here too? Eg empty string, etc

**Review by nyalldawson — APPROVED** [review]

**Comment by manisandro** [comment]
Thanks for the review

**Pull request merged**